### PR TITLE
feat(Queue): rework battleground queue system

### LINKED ISSUE #12: Join as a Group doesn't work [$100]
Joining as a group (party) does not work.

It would be nice if it can work and somehow try (but of course cannot guarantee) to maintain the original group members within the same team.

EDIT: there is a bounty on this issue. To consider it fixed, it MUST be properly tested and ALL corner cases should work. Including all different configurations enabled, for example:

```
CFBG.BalancedTeams = 1
CFBG.Include.Avg.Ilvl.Enable = 1
CFBG.Players.Count.In.Group = 2
CFBG.EvenTeams.Enabled = 1
CFBG.EvenTeams.MaxPlayersThreshold = 5
```

<bountysource-plugin>

---
There is a **[$100 open bounty](https://www.bountysource.com/issues/91199193-join-as-a-group-doesn-t-work?utm_campaign=plugin&utm_content=tracker%2F146727645&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146727645&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes:
- Add support new queue system from PR https://github.com/azerothcore/azerothcore-wotlk/pull/10817
- Add support groups
- Now players are now mixed in queue, not when entering in bg

## How to test:
- Need test with all config options
- Need test full balance system (low level and other)

## Issues:
- Close #12

### CODE REVIEW

File: src/CFBG.cpp [hunk 1]
```diff
@@ -154,141 +141,67 @@ TeamId CFBG::GetLowerTeamIdInBG(Battleground* bg, Player* player)
...
-                        }
-                        // no else, if the player has more ilvl than the queue, put him on the weaker team.
-                    }
-                }
+            if (std::abs(avgLvlAlliance - avgLvlHorde) >= 0.5f)
+            {
+                if (avgLvlAlliance < avgLvlHorde)
+                    team = TEAM_ALLIANCE;
+                else
+                    team = TEAM_HORDE;
```
Comment (Helias) [inline src/CFBG.cpp]:
```suggestion
                team = avgLvlAlliance < avgLvlHorde ? TEAM_ALLIANCE : TEAM_HORDE;
```

**Review by Helias — APPROVED** [review]

**Comment by Helias** [comment]
for me the PR is ok, hope you tested scenarios like this https://github.com/azerothcore/mod-cfbg/pull/86#issuecomment-991655228

**Comment by Nefertumm** [comment]
Conflicts

**Comment by GoMateoGo** [comment]
1. I set 2 players to start the battlefield

2.2 players have joined the queue and display 2 / 2

3. You should enter the battlefield normally at this time But no

4. Unless there is a third player in the queue, display 3 / 2 At this time, there was still no dialogue on entering the battlefield

5. Unless one of the three players leaves the queue, the other two players will pop up the enter battlefield dialogue at this moment

**Comment by GoMateoGo** [comment]
CFBG.Enable = 1
CFBG.BalancedTeams = 1
CFBG.BalancedTeams.Class.LowLevel = 0
CFBG.BalancedTeams.Class.MinLevel = 10
CFBG.BalancedTeams.Class.MaxLevel = 19
CFBG.BalancedTeams.Class.LevelDiff = 2
CFBG.Include.Avg.Ilvl.Enable = 0
CFBG.Players.Count.In.Group = 3
CFBG.EvenTeams.Enabled = 0
CFBG.EvenTeams.MaxPlayersThreshold = 5
CFBG.ResetCooldowns = 1
CFBG.Show.PlayerName = 1

**Comment by GoMateoGo** [comment]
Supplement:

The dialog box for entering the battlefield cannot be popped up no matter how many people are involved

Unless any one in the queue leaves the queue, the enter battlefield dialog box will pop up for others in the queue

**Pull request merged**